dashboard/app/models/user.rb

```diff
@@ -2261,6 +2261,41 @@ def providers
     end
   end
 
+  # Returns number of times a user has attempted to join a section in the last 24 hours
+  # Returns nil if no section join attempts
+  def num_section_attempts
+    properties['section_attempts'] || 0
+  end
+
+  # There aare two possible states in which we would want to reset section attempts
+  # 1) Initialize for the first time 2) 24 hours have passed since last attempt
+  def reset_section_attempts?
+    !num_section_attempts || !properties['section_attempts_last_reset'] || num_section_attempts == 0 || (DateTime.now - DateTime.parse(properties['section_attempts_last_reset'])).to_i > 0
+  end
+
+  def reset_section_attempts
+    properties['section_attempts'] = 0
+    properties['section_attempts_last_reset'] = DateTime.now.to_s
+    save
+  end
+
+  def reset_section_attempts_if_needed
+    # Failed section attempts reset after 24 hours
+    reset_section_attempts if reset_section_attempts?
+  end
+
+  def display_captcha?
+    reset_section_attempts_if_needed
+    num_section_attempts >= 3
+  end
+
+  # Let's make sure we only hit this code when the user is already registered
```

Comment:
this comment is a little confusing to me, because it sounds like you're about to do that (make sure user is registered), but the code that follows doesn't appear to do it.

are we accomplishing that by checking for `current_user` before calling this? (i'm not sure if having a `current_user` means they are registered). if that's the case, it seems like the fact that this is an instance method means it could only be called on a registered user, in which case i'd just remove this comment.

if we're verifying that the use is registered somewhere else, i'd suggest either moving that verification into this meothd, or at least explaining how/where we're doing it in this comment.

Reply:
I think this was a comment that I left to myself but did not remove. Should have labeled it with `TODO` — the issue last time was calling methods in the user model on `@user` instances that had not yet been persisted to the database, so I wanted to be extra careful that doesn’t happen again by tracing all the method calls in which `save` Is invoked.

Reply:
Also would love to get confirmation that `current_user` does in fact always reference a User object that has already been persisted to the database. The tricky thing about Rails is that instance methods can be called on objects before they ever go into the database. Only a `save` call fire of the SQL necessary to persist any changes.

Reply:
I'm not sure if `current_user` is always persisted but you could always check [`current_user.persisted?`](https://apidock.com/rails/ActiveRecord/Persistence/persisted%3F)

Reply:
Started a discussion about the `current_user` method inherited from `ApplicationController` in this [Slack thread](https://codedotorg.slack.com/archives/C0T10HG6N/p1611198096002600).